lib/shared/sdk/usbboot/protocol.js

```diff
@@ -0,0 +1,314 @@
+
+/*
+ * This work is heavily based on https://github.com/raspberrypi/usbboot
+ * Copyright 2016 Raspberry Pi Foundation
+ */
+
+'use strict'
+
+const _ = require('lodash')
+const Bluebird = require('bluebird')
+const usb = require('./usb')
+
+// The equivalent of a NULL buffer, given that node-usb complains
+// if the daata argument is not an instance of Buffer
+const NULL_BUFFER_SIZE = 0
+const NULL_BUFFER = Buffer.alloc(NULL_BUFFER_SIZE)
+
+/**
+ * @summary The size of the boot message bootcode length section
+ * @type {Number}
+ * @constant
+ */
+const BOOT_MESSAGE_BOOTCODE_LENGTH_SIZE = 4
+
+/**
+ * @summary The offset of the boot message bootcode length section
+ * @type {Number}
+ * @constant
+ */
+const BOOT_MESSAGE_BOOTCODE_LENGTH_OFFSET = 0
+
+/**
+ * @summary The size of the boot message signature section
+ * @type {Number}
+ * @constant
+ */
+const BOOT_MESSAGE_SIGNATURE_SIZE = 20
+
+/**
+ * @summary The offset of the file message command section
+ * @type {Number}
+ * @constant
+ */
+const FILE_MESSAGE_COMMAND_OFFSET = 0
+
+/**
+ * @summary The size of the file message command section
+ * @type {Number}
+ * @constant
+ */
+const FILE_MESSAGE_COMMAND_LENGTH = 4
+
+/**
+ * @summary The offset of the file message file name section
+ * @type {Number}
+ * @constant
+ */
+const FILE_MESSAGE_FILE_NAME_OFFSET = FILE_MESSAGE_COMMAND_LENGTH
+
+/**
+ * @summary The size of the file message file name section
+ * @type {Number}
+ * @constant
+ */
+const FILE_MESSAGE_FILE_NAME_LENGTH = 256
+
+/**
+ * @summary The GET_STATUS usb control transfer request code
+ * @type {Number}
+ * @constant
+ * @description
+ * See http://www.jungo.com/st/support/documentation/windriver/811/wdusb_man_mhtml/node55.html#usb_standard_dev_req_codes
+ */
+const USB_REQUEST_CODE_GET_STATUS = 0
+
+/**
+ * @summary The maximum buffer length of a usbboot message
+ * @type {Number}
+ * @constant
+ */
+const USBBOOT_MESSAGE_MAX_BUFFER_LENGTH = 0xffff
+
+/**
+ * @summary The delay to wait between each USB read/write operation
+ * @type {Number}
+ * @constant
+ * @description
+ * The USB bus seems to hang if we execute many operations at
+ * the same time.
+ */
+const USB_REQUEST_DELAY_MS = 1000
+
+/**
+ * @summary The timeout for USB bulk transfers, in milliseconds
+ * @type {Number}
+ * @constant
+ */
+const USB_BULK_TRANSFER_TIMEOUT_MS = 1000
+
+/**
+ * @summary The size of the usbboot file message
+ * @type {Number}
+ * @constant
+ */
+exports.FILE_MESSAGE_LENGTH = FILE_MESSAGE_COMMAND_LENGTH + FILE_MESSAGE_FILE_NAME_LENGTH
+
+/**
+ * @summary The "get file size" file message command name
+ * @type {String}
+ * @constant
+ */
+exports.FILE_MESSAGE_COMMAND_GET_FILE_SIZE = 'GetFileSize'
+
+/**
+ * @summary The "read file" file message command name
+ * @type {String}
+ * @constant
+ */
+exports.FILE_MESSAGE_COMMAND_READ_FILE = 'ReadFile'
+
+/**
+ * @summary The "done" file message command name
+ * @type {String}
+ * @constant
+ */
+exports.FILE_MESSAGE_COMMAND_DONE = 'Done'
```

Comment:
Instead of 3 separate constants, I guess you could do:
```javascript
const FILE_MESSAGE_COMMANDS = {
    GET_FILE_SIZE: 'GetFileSize',
    READ_FILE: 'ReadFile',
    DONE: 'Done'
}
```
?

Reply:
Done!